src/services/mapCode.ts

```diff
@@ -0,0 +1,484 @@
+                    /*setParentNodes*/ true,
+                ),
+            body: (cw: SourceFile) => (cw.statements[0] as ClassLikeDeclaration).members,
+        },
+    ];
+    const parsedNodes = [];
+    for (const { parse, body } of nodeKinds) {
+        const sourceFile = parse();
+        const bod = body(sourceFile);
+        if (bod.length && sourceFile.parseDiagnostics.length === 0) {
+            // If we run into a case with no parse errors, this is likely the right kind.
+            return bod;
+        }
+        // We only want to keep the ones that have some kind of body.
+        else if (bod.length) {
+            // Otherwise, we'll need to look at others.
+            parsedNodes.push({ sourceFile, body: bod });
+        }
+    }
+    // Heuristic: fewer errors = more likely to be the right kind.
+    const { body } = parsedNodes.sort(
+        (a, b) =>
+            a.sourceFile.parseDiagnostics.length -
+            b.sourceFile.parseDiagnostics.length,
+    )[0];
+    return body;
+}
+
+function placeNodeGroup(
+    originalFile: SourceFile,
+    changeTracker: ChangeTracker,
+    changes: NodeArray<Node>,
+    focusLocations?: TextSpan[],
+) {
+    // 1. Grab the first node, see what kind it is, to see whether we'll need
+    //    to expand our scope.
+    // 2. Get nodes for each focusLocation by finding the specific node that
+    //    contains it, and making a comparison with the first node type.
+    // 3. Figure out how much of a range we're going to replace or insert by
+    //    using first and last node from `nodes`. Use node types and names to
+    //    match ranges.
+    if (isClassElement(changes[0]) || isTypeElement(changes[0])) {
+        placeClassNodeGroup(
+            originalFile,
+            changeTracker,
+            changes as NodeArray<ClassElement>,
+            focusLocations,
+        );
+    }
+    else {
+        placeStatements(
+            originalFile,
+            changeTracker,
+            changes as NodeArray<Statement>,
+            focusLocations,
+        );
+    }
+}
+
+function placeClassNodeGroup(
+    originalFile: SourceFile,
+    changeTracker: ChangeTracker,
+    changes: NodeArray<ClassElement> | NodeArray<TypeElement>,
+    focusLocations?: TextSpan[],
+) {
+    // We have one or more class-ish members.
+    //
+    // 1. If focusLocations is null/undefined, find the first class in the
+    //    SourceFile and insert/replace into that class.
+    // 2. If we have focusLocations, go over each one and:
+    //   a. Find the nearest class or interface declaration that contains the focusLocation.
+    let classOrInterface:
+        | ClassLikeDeclaration
+        | InterfaceDeclaration
+        | undefined;
+    if (!focusLocations || !focusLocations.length) {
+        classOrInterface = originalFile.statements.find(
+            stmt => isClassLike(stmt) || isInterfaceDeclaration(stmt),
+        ) as ClassLikeDeclaration | InterfaceDeclaration | undefined;

```

Comment:
I couldn't find what you're referring to?

Reply:
I found it in corePublic.ts:162:

`export function find<T, U extends T>(array: readonly T[] | undefined, predicate: (element: T, index: number) => element is U, startIndex?: number): U | undefined;`

I misread this as using that function instead of Array.prototype.find. But, using `find` would mean that `classOrInterface = find(originalFile.statements, or(isClassLike, isInterfaceDeclaration));` would work